content/blog/typescript-type-generics/index.md

```diff
@@ -0,0 +1,340 @@
+
+# The Problem {#generic-usecase-setup}
+
+Type generics — on the highest level — _allow you to accept arbitrary data instead of strict typing, making it possible to broaden a type's scope_.
+
+For example, what if you wanted to make a function that took an argument of `returnProp` and returned the `returnProp` value itself ([the formal name for a function like this is an **identity function**](https://en.wikipedia.org/wiki/Identity_function))? Without type generics, providing a typing for a function like this could be difficult.
+
+Take the following implementation and consider its limitations:
+
+```typescript
+function returnProp(returnProp: string): string {
+	return returnProp;
+}
+
+returnProp('Test'); // ✅ This is fine
+returnProp(4); // ❌ This would fail as `4` is not a string
+```
+
+In this case, we want to make sure that every possible input type is available for the prop type. Let's take a look at a few potential solutions, with their various pros and cons, and see if we can find a solution that fits the requirements for providing typing for a function like this.
+
+## Potential Solution 1: Unions {#generic-usecase-setup-union-solution}
+
+One potential solution to this problem might be TypeScript unions. _Unions allow us to define an `or` condition of sorts for our types_. As we want to allow various types for inputs and outputs, perhaps that can help us here!
+
+Using this method, if we wanted to accept numbers, we could add that as a union:
+
+```typescript
+function returnProp(returnProp: string | number): string | number {
+	return returnProp;
+}
+
+returnProp('Test'); // ✅ This is fine
+const shouldBeNumber = returnProp(4); // ✅ This won't show errors now
+```
+
+However, unions have some limitations. You'll find that this doesn't give the example you might want:
+
+```typescript
+// ❌ This will yield an error
+// > Operator '+' cannot be applied to types '4' and 'string | number'.
+const newNumber = shouldBeNumber + 4;
+```
+
+The reason that the operation `shouldBeNumber + 4` yields this error is because it doesn't know which type to give you given your input, since the output is also explicitly a union (meaning that it could be a string OR a number returned — one of them supporting numeric operations, one of them not).
+
+### Potential Solutions Disclaimer {#silly-examples-disclaimer}
+
+> Author's note:
+>
+> If you were using unions in your property definitions and left your return type blank, TypeScript would be able to infer what the return type should be just fine.
+>
+> That said, we're trying to build on concepts, so we're trying to provide some examples of where this might be used and what it does. There are also instances, such as type definition files, where this inference might not be available to an author of typings, as well as other limitations with this method that we'll see later,
```

Comment:
```suggestion
> That said, we're trying to build on concepts, so we're trying to provide some examples of where this might be used and what it does. There are also instances, such as type definition files, where this inference might not be available to an author of typings, as well as other limitations with this method that we'll see later.
```